src/axom/core/Array.hpp

```diff
@@ -55,11 +87,58 @@ bool operator==(const Array<T>& lhs, const Array<T>& rhs);
  * \return true if the Arrays do not have the same allocator ID, are not of
  * equal length, or do not have the same elements.
  */
-template <typename T>
-bool operator!=(const Array<T>& lhs, const Array<T>& rhs);
+template <typename T, IndexType dim>
+bool operator!=(const Array<T, dim>& lhs, const Array<T, dim>& rhs);
 
 /// @}
 
+// Stuff we only want to make available to the 1D case (where we want to closely emulate std::vector)
+template <typename T, IndexType dim, typename ArrayType>
+struct Only1D
+{ };
+
+template <typename T, typename ArrayType>
+struct Only1D<T, 1, ArrayType>
+{
+  IndexType size() const { return asDerived().fullSize(); }
+
+  /*!
+     * \brief Push a value to the back of the array.
+     *
+     * \param [in] value the value to the back.
+     *
+     * \note Reallocation is done if the new size will exceed the capacity.
+     */
+  void push_back(const T& value);
+
+  /*!
+     * \brief Push a value to the back of the array.
+     *
+     * \param [in] value the value to move to the back.
+     *
+     * \note Reallocation is done if the new size will exceed the capacity.
+     */
+  void push_back(T&& value);
+
+  /*!
+     * \brief Inserts new element at the end of the Array.
+     *
+     * \param [in] args the arguments to forward to constructor of the element.
+     *
+     * \note Reallocation is done if the new size will exceed the capacity.
+     * \note The size increases by 1.
+     */
+  template <typename... Args>
+  void emplace_back(Args&&... args);
```

Comment:
As we discussed earlier, these may be difficult to implement on a device. That being said, this array class may not have to be usable directly on a device-- cuda kernels would likely only ever need an ArrayView